core-plugins/src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java

```diff
@@ -0,0 +1,376 @@
+    }
+    return filedMappings;
+  }
+
+  /**
+   * Add field to json element
+   *
+   * @param rootPath    absolute path of field
+   * @param path        relative path of the field
+   * @param jsonElement current json element
+   * @return updated json element
+   */
+  private JsonElement addField(List<String> rootPath, List<String> path, JsonElement jsonElement) {
+    if (jsonElement.isJsonObject()) {
+      if (!jsonElement.getAsJsonObject().has(path.get(0))) {
+        if (path.size() > 1) {
+          jsonElement.getAsJsonObject().add(path.get(0), addField(rootPath, path.subList(1, path.size()),
+                                                                  GSON.fromJson("{}", JsonElement.class)));
+        } else {
+          jsonElement.getAsJsonObject().add(path.get(0), GSON.toJsonTree(rootPath));
+        }
+      }
+    }
+    return jsonElement;
+  }
+
+  /**
+   * Return field mapping with/without unmapped fields based on config
+   *
+   * @param inputSchema {@link Schema} current input schema
+   * @return {@link JsonElement} field mapping
+   */
+  private JsonElement getFinalFieldMappingJson(Schema inputSchema) {
+    boolean includeNonMappedFields = createRecordTransformConfig.getIncludeNonMappedFields().equals("on");
+    final JsonElement fieldMappingJson = createRecordTransformConfig.getFieldMappingJson();
+    if (includeNonMappedFields) {
+      final List<List<String>> inputSchemaMap = generateFieldMapForInputSchema(new ArrayList<>(), inputSchema);
+      final List<List<String>> fieldMapToList = fieldMapToList(fieldMappingJson);
+      final List<List<String>> difference = differenceBetweenMaps(inputSchemaMap, fieldMapToList);
+      return includeMissingFields(fieldMappingJson, difference);
+    }
+    return fieldMappingJson;
+  }
+
+  /**
+   * Generate output schema
+   */
+  private Schema getOutputSchema(Schema inputSchema, FailureCollector collector) {
+    final JsonElement fieldMappingJson = getFinalFieldMappingJson(inputSchema);
+    if (inputSchema == null) {
+      collector.addFailure("Missing input schema.", "Please provide valid input schema.");
+    }
+    if (fieldMappingJson.isJsonNull()) {
+      collector.addFailure("Empty mapping field.", "Please provide valid mapping field.")
+        .withConfigProperty(CreateRecordTransformConfig.FIELD_MAPPING);
+    }
+    collector.getOrThrowException();
+    final Schema schema = generateFields(inputSchema, collector, "record",
+                                         fieldMappingJson.getAsJsonObject());
+    return schema;
+  }
+
+  /**
+   * Get field from structured record
+   *
+   * @param structuredRecord {@link StructuredRecord} record to read the data from
+   * @param pathMap          {@link JsonArray} path of the field to read from
+   * @return {@link StructuredRecord} property containing the value
+   */
+  private Object getField(StructuredRecord structuredRecord, JsonArray pathMap) {
+    if (pathMap.size() == 1) {
```

Comment:
why do we need a special case for pathMap.size() == 1? Can't it be handled in the while loop below?